### PR TITLE
Wrong dependency resolution plus performance optimizations when building

### PR DESCRIPTION
**Pull request by freddydk**
Two new parameter in Sort-AppFoldersByDependencies - onlyTheseAppFoldersPlusDepending and SkipApps. If specified, then only the appFolders specified in the first array plus the apps depending on these recursively will be sorted. The remaining folder not included in the sorted appfolders list will be returned in skipApps. This is used for incrementals builds in AL-Go for GitHub.

Remove some excessive logging

If Import-TestToolkitToBcContainer was running with useCompilerFolder, then it would require a container to be present at that time.

Make Run-AlPipeline do just-in-time creation of CompilerFolder and Containers - meaning that if you are using useCompilerFolder and do not have doNotPublishApps set - it will create a compilerFolder and compile all apps using that, and afterwards - it will create the container, publish the apps and run the tests.

Also...
Fixes https://github.com/microsoft/AL-Go/issues/1338

### CODE REVIEW

**Review by mazhelez — COMMENTED** [review]
Left some comments. I'll circle back to this PR later on.

File: AppHandling/Run-AlPipeline.ps1 [hunk 1]
```diff
@@ -1769,6 +1874,41 @@ Write-Host -ForegroundColor Yellow @'
...
+
+    $prebuiltAppFileName = ''
+    $prebuiltAppName = "$($appJson.Publisher)_$($appJson.Name)".Split([System.IO.Path]::GetInvalidFileNameChars()) -join ''
+    if ($buildArtifactFolder) {
+        $prebuiltFolderName = Join-Path $buildArtifactFolder "$(if($app){"Apps"}else{"TestApps"})"
+        $prebuiltAppFileName = Join-Path $prebuiltFolderName "$($prebuiltAppName)_*.*.*.*.app"
+        if (Test-Path $prebuiltAppFileName) {
+            $prebuiltAppFileName = (Get-Item $prebuiltAppFileName).FullName
+            if ($prebuiltAppFileName -is [Array]) {
+                Write-Host "Multiple apps found for prebuilt app - rebuilding app!"
```
[the remark this answers is not in the document]
Comment (freddydk) [reply AppHandling/Run-AlPipeline.ps1:1889]: That - or app names which hits this strange pattern, like:

MyApp_1.2.3.4
MyApp_Number2_1.2.3.4

The second app would match "MyApp_*.*.*.*" - and here, I would just decide to rebuild and throw in a comment to explain why.

File: AppHandling/Sort-AppFoldersByDependencies.ps1 [hunk 2]
```diff
@@ -84,50 +98,72 @@ try {
     $script:unresolvedDependencies = $()
 
     function AddAnApp { Param($anApp) 
+        $rebuildThis = $false
```
Comment (mazhelez) [inline AppHandling/Sort-AppFoldersByDependencies.ps1:101]: I don't get why this is called `$rebuildThis`. Nothing is being built around here.

**Review by mazhelez — DISMISSED** [review]
Looks okay. It feels like the sorting functionality is getting overly complicated. 
Some tests will really showcase the usage of the new parameter, but I didn't find existing tests.

**Review by mazhelez — APPROVED** [review]

**Pull request merged**